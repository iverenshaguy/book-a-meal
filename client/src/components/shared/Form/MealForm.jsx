import React, { Fragment } from 'react';
import { RenderInput, RenderFileInput } from '../FormComponents';
import { formPropTypes, mealObjPropTypes } from '../../../helpers/proptypes';

/**
 * @exports
 * @function MealForm
 * @param {object} props
 * @returns {JSX} MealForm
 */
const MealForm = (props) => {
  const {
    state, handlers, meal, updating
  } = props;

  return (
    <Fragment>
      <RenderInput
        type="text"
        name="title"
        id="title"
        label="Meal Name"
        required
        value={state.values.title}
        placeholder=""
        handleChange={handlers.handleChange}
        handleBlur={handlers.handleBlur}
        handleFocus={handlers.handleFocus}
        meta={{
          touched: state.touched.title,
          error: state.error.title,
          asyncValidating: state.asyncValidating
        }}
      />
      <RenderInput
        type="number"
        name="price"
        id="price"
        label="Price"
        required
        value={state.values.price}
        placeholder=""
        handleChange={handlers.handleChange}
        handleBlur={handlers.handleBlur}
        handleFocus={handlers.handleFocus}
        meta={{
          touched: state.touched.price,
          error: state.error.price
        }}
      />
      <RenderFileInput
        id="imageUrl"
        name="imageUrl"
        label="Meal Image"
        updating={updating}
        formerImgUrl={meal ? meal.imageUrl : 'https://res.cloudinary.com/iverenshaguy/image/upload/v1532540264/bookameal/default-img.jpg'}
        successCallBack={imageUrl => handlers.handleChange({ target: { name: 'imageUrl', value: imageUrl } })}
      />
      <RenderInput
        type="textarea"
        name="description"
        label="Description (not more than 50 characters)"
        id="description"
        rows={2}
        maxLength={100}
        required={false}
        value={state.values.description}
        placeholder=""
        handleChange={handlers.handleChange}
        handleBlur={handlers.handleBlur}
        handleFocus={handlers.handleFocus}
        meta={{
          touched: state.touched.description,
          error: state.error.description
        }}
      />
      <RenderInput
        type="checkbox"
        name="vegetarian"
        label="Suitable for Vegetarians"
        id="vegetarian"
        required={false}
        value={state.values.vegetarian}
        placeholder=""
        handleChange={handlers.handleChange}
        handleBlur={handlers.handleBlur}
        handleFocus={handlers.handleFocus}
        meta={{
          touched: state.touched.vegetarian,
          error: state.error.vegetarian
        }}
      />
    </Fragment>
  );
};

MealForm.propTypes = {
  ...formPropTypes('addMeal'),
  meal: mealObjPropTypes
};

MealForm.defaultProps = {
  meal: null
};

export default MealForm;
