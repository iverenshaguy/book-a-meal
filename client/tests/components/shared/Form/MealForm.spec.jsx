import React from 'react';
import MealForm from '../../../../src/components/shared/Form/MealForm';
import { formComponentSetup } from '../../../setup/formSetup';
import { mealsObj } from '../../../setup/mockData';

describe('MealForm', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should render the Meal Form correctly when type prop is addMeal', () => {
    const { handlers, state } = formComponentSetup('addMeal');
    const wrapper = shallow(<MealForm type="addMeal" state={state} handlers={handlers} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render the Meal Form correctly when type prop is editMeal', () => {
    const { handlers, state } = formComponentSetup('editMeal');
    const wrapper = shallow(<MealForm type="editMeal" state={state} handlers={handlers} meal={mealsObj.meals[0]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
