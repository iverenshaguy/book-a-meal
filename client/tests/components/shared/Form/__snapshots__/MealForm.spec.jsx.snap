// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealForm should render the Meal Form correctly when type prop is addMeal 1`] = `
<Fragment>
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="title"
    label="Meal Name"
    meta={
      Object {
        "asyncValidating": false,
        "error": null,
        "touched": false,
      }
    }
    name="title"
    placeholder=""
    required={true}
    type="text"
    value=""
  />
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="price"
    label="Price"
    meta={
      Object {
        "error": null,
        "touched": false,
      }
    }
    name="price"
    placeholder=""
    required={true}
    type="number"
    value=""
  />
  <Connect(RenderFileInput)
    formerImgUrl="https://res.cloudinary.com/iverenshaguy/image/upload/v1532540264/bookameal/default-img.jpg"
    id="imageUrl"
    label="Meal Image"
    name="imageUrl"
    successCallBack={[Function]}
  />
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="description"
    label="Description (not more than 50 characters)"
    maxLength={100}
    meta={
      Object {
        "error": null,
        "touched": false,
      }
    }
    name="description"
    placeholder=""
    required={false}
    rows={2}
    type="textarea"
    value=""
  />
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="vegetarian"
    label="Suitable for Vegetarians"
    meta={
      Object {
        "error": null,
        "touched": false,
      }
    }
    name="vegetarian"
    placeholder=""
    required={false}
    type="checkbox"
    value=""
  />
</Fragment>
`;

exports[`MealForm should render the Meal Form correctly when type prop is editMeal 1`] = `
<Fragment>
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="title"
    label="Meal Name"
    meta={
      Object {
        "asyncValidating": false,
        "error": null,
        "touched": false,
      }
    }
    name="title"
    placeholder=""
    required={true}
    type="text"
    value="Jollof Rice and Chicken"
  />
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="price"
    label="Price"
    meta={
      Object {
        "error": null,
        "touched": false,
      }
    }
    name="price"
    placeholder=""
    required={true}
    type="number"
    value="2500.00"
  />
  <Connect(RenderFileInput)
    formerImgUrl="http://www.preciouscore.com/wp-content/uploads/2017/11/How-to-cook-jollof-rice-in-the-oven-750x500.jpg"
    id="imageUrl"
    label="Meal Image"
    name="imageUrl"
    successCallBack={[Function]}
  />
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="description"
    label="Description (not more than 50 characters)"
    maxLength={100}
    meta={
      Object {
        "error": null,
        "touched": false,
      }
    }
    name="description"
    placeholder=""
    required={false}
    rows={2}
    type="textarea"
    value=""
  />
  <RenderInput
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleFocus={[MockFunction]}
    id="vegetarian"
    label="Suitable for Vegetarians"
    meta={
      Object {
        "error": null,
        "touched": false,
      }
    }
    name="vegetarian"
    placeholder=""
    required={false}
    type="checkbox"
    value={false}
  />
</Fragment>
`;
